### DISCUSSION TITLE
Does toga support multiprocessing?

### DISCUSSION THREAD

**Discussion by Jzhenli** [beeware/toga#2754] [Q&A] [answered]
Hi, I use the multiprocessing in toga app for some heavy cpu loading tasks, it can run well with "briefcase dev", but when use the "briefcase run" will open multiple instances of the app, is this a bug? if yes, how to fix it? BTW, my system OS is win11.

**Comment by freakboy3742** [comment, marked as answer]
As noted in [this ticket](https://github.com/beeware/briefcase-windows-VisualStudio-template/issues/8), multiprocessing isn't currently possible.

**Comment by Jzhenli** [comment]
@freakboy3742 Seems the windows platform use multiprocessing as spawn mode, and the stub not support it. Now I want to know if the android platform support multiprocessing? Android based on linux and the multiprocessing may be default as fork mode, right?

**Reply by freakboy3742** [reply]
No. Android and iOS both prohibit the use of fork and spawn. Although they are both "unix-like" operating systems, their process model largely assumes that the app is a single process, running in the foreground. Calls to fork/spawn will either fail, or cause lockups in the app.

**Reply by Jzhenli** [reply]
I originally wanted to use this solution: togo webview as ui, and create a process to run fastapi as a background server, then packaged them as an android app, so it's not possible now. Would u have any suggestion for integrate toga webview and fastapi web server? @freakboy3742

**Reply by freakboy3742** [reply]
You use threads, not processes. This is what the [positron-static](https://github.com/beeware/toga/tree/main/examples/positron-static) and [positron-django](https://github.com/beeware/toga/tree/main/examples/positron-django) examples do to achieve exactly what you've described - a web view-based GUI backed by a web server on the same app.

I haven't got a "ready to use" example for FastAPI - but the general approach should be exactly the same.

**Reply by Jzhenli** [reply]
FastAPI is primarily an ASGI framework, and the two examples you offered are WSGI framework, they have some differences, I don't know how to stop ASGI  in the thread.

**Reply by freakboy3742** [reply]
You'll need to take that up with the FastAPI folks. I've never used FastAPI, so I can't advise on how to run a FastAPI server.

However, if there's an async end point that you can `await` that runs a server, you may not need a thread at all - you may be able to use `asyncio.create_task()` to install the async handler directly into the Toga app's GUI event loop.

**Reply by michealroberts** [reply]
@freakboy3742 Integrating an ASGI server (like FastAPI) directly into Toga’s GUI loop with asyncio.create_task() usually won't work because Toga does not natively run on an asyncio event loop. Toga’s main loop is an event loop provided by the underlying OS GUI toolkit. There’s no built-in, official mechanism to “merge” Toga’s native event loop with asyncio in a single thread. 

As a result, running the FastAPI server in a separate thread (or process) will have to be the standard approach.

**Reply by freakboy3742** [reply]
> @freakboy3742 Integrating an ASGI server (like FastAPI) directly into Toga’s GUI loop with asyncio.create_task() usually won't work because Toga does not natively run on an asyncio event loop. 

@michealroberts I don't know what has given you that impression, but your statement is **categorically** incorrect. I can point you at *dozens* of examples in the Toga codebase that prove that Toga's provides a fully featured asyncio event loop that is fully integrated with the platform's GUI event loop. I have *personally* spent many days implementing and testing the specific feature that you claim doesn't exist. As a specific example, `asyncio.create_task()` works *fine*. If you want proof of this, [try the `handlers` example app](https://github.com/beeware/toga/blob/main/examples/handlers/handlers/app.py#L77).

You have logged #3325 describing that you're having difficulty with running an ASGI server. The code you've presented there doesn't use any of the asyncio APIs that FastAPI provides - it's using `uvicorn.run()` *in a thread*. As I indicated on Discord last night, the stack traces you've provided suggest the issue is one of *configuring the FastAPI server*, not anything to do with Toga supporting asyncio; and you probably *don't* want to be using `uvicorn.run()` - you will almost certainly have more luck with using one of the [lower level examples of server usage that are described *on the uvicorn homepage*](https://www.uvicorn.org/#config-and-server-instances) - in particular, the one that *runs the server in an existing asyncio loop*.

I'll take a closer look at the specific example on that ticket when I get a chance - It's currently Saturday morning for me, and I'm travelling for work next week.
